### ISSUE TITLE
CSI ephemeral volumes: idempotent NodeUnpublishVolume

### ISSUE THREAD

**Issue by pohly** [kubernetes/kubernetes#90752]
**What happened**:

During testing with PMEM-CSI, `NodeUnpublishVolume` was called twice by kubelet, presumably because the first call took too long (deleting volumes does take time). The second call blocked, waiting for the first to complete its work, and then returned `NOT_FOUND` because the first call had removed all information about the volume. That caused kubelet to keep calling `NodeUnpublishVolume` forever and the pod never got deleted.

**What you expected to happen**:

Returning `NOT_FOUND` is required by the CSI spec. The driver cannot know that the volume was ephemeral because it now longer has any information about it. The csi-driver-host-path does the same: https://github.com/kubernetes-csi/csi-driver-host-path/blob/941f7a8aa7b991c11293e806de80e6b6483d3d01/pkg/hostpath/nodeserver.go#L203-L206

Perhaps kubelet then needs to be changed and accept that `NOT_FOUND` indicates success? But doing that for persistent volumes would be a spec violation, too, so it has to be limited to ephemeral volumes.

/cc @jsafrane for suggestions...

**How to reproduce it (as minimally and precisely as possible)**:

We would have to artificially slow down `NodeUnpublishVolume` in a driver to trigger the error condition in kubelet.

/sig storage

**Comment by jsafrane** [comment]
Is it similar to https://github.com/container-storage-interface/spec/pull/375 ?

It affects also non-ephemeral volumes - when NodeUnpublishVolume times out from whatever reason, it may succeed in the background and something may delete even a non-ephemeral volume. Kubelet then re-tries to delete the volume indefinitely.

IMO, if SP can't find the volume and at the same time it can check that such volume is not mounted, e.g. by checking that corresponding directory is not a mountpoint,  NodeUnpublish and NodeUnstage should return OK - the volume has been unpublished / unstaged.

**Comment by pohly** [comment]
Yes, that looks very similar. #375 was about the controller level, but the node level has the same issue.
